cip/1.accepted/CIP2017-01-18-isomorphic-matching.adoc

```diff
@@ -0,0 +1,79 @@
+= CIP2017-01-18 - Isomorphic Matching Semantics
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Stefan Plantikow <stefan.plantikow@neotechnology.com>
+
+This proposal is a response to CIR-2017-174.
+
+=== Proposal: Add new uniqueness modes
+
+It is proposed to add the capability to select one of three uniqueness modes for a uniqueness scope:
+
+* `MATCH ALL`: Impose no uniqueness requirements on candidate matches
+* `MATCH UNIQUE RELATIONSHIPS`: Only consider candidate matches that are relationship-unique
+* `MATCH UNIQUE NODES`: Only consider candidate matches that are node-unique
+
+The default uniqueness mode used by `MATCH` (without a further specification of the preferred uniqueness mode) is relationship-unique matching.
+
+`MATCH ALL` does not reject any paths - not even paths containing cycles - and hence can lead to infinite result sets for the whole query.
+It is recommended that implementations generate at least a warning when static analysis is not able to proof query termination due to the chosen uniqueness mode.
+
+It is conceivable that this approach for the specification of uniqueness is extensible by adding further ways to restrict uniqueness.
+
+=== Proposal: Specifying the uniqueness mode of a subquery
+
+Changing the uniqueness mode of a sub query recursively changes the default uniqueness mode for all contained `MATCH` clauses unless it is overridden again. Examples:
+
+* `MATCH <uniqueness-modes> { MATCH ... } ...`
+* `DO <uniqueness-modes> { MATCH ... } ...`
+
+=== Proposal: Default uniqueness mode
+
+Additionally, it is proposed that a conforming implementation should provide a pre-parser option for defining a default uniqueness level for use with regular pattern matching.
```

Comment:
I'm not convinced this kind of recommendation belongs in a CIP. Is it not well understood that an implementing system would provide ways of changing defaults?